### PR TITLE
Support video input with data_id mechanism

### PR DESCRIPTION
**Pull request by banasraf**
This PR adds support for `data_id` mechanism enabling decoupled models for video input.

Decoupled model in Triton is a model that can send 0...n responses for each request (as opposed to standard models which have 1 to 1 response-request relationship).
In case of DALI it lets us utilize video input by sending multiple responses for each sent video file.
This is achieved by using the data_id mechanism - it tells us when the video file has been consumed, so we know when to stop processing the current request and accept the next one.

The most important changes are those in dali_executor/** files. The clue of this feature is to set the `data_id` for each request and then correctly determine if the input was consumed.

### CODE REVIEW

File: qa/L0_video_input_decoupled/client.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            outp = t_client.InferRequestedOutput('OUTPUT')
+
+            request_id = str(req_id)
+            triton_client.async_stream_infer(model_name=model_name,
+                                             inputs=[inp],
+                                             request_id=request_id,
+                                             outputs=[outp])
+
+
+            expected_result = eager.experimental.decoders.video([input_data])
```
Comment (JanuszL) [inline qa/L0_video_input_decoupled/client.py:92]:
```suggestion
            eager.experimental.decoders.video([input_data])
```

File: src/dali_executor/dali_executor.cc [hunk 2]
```diff
@@ -148,7 +154,12 @@ std::vector<OutputInfo> DaliExecutor::Run(const std::vector<IDescr>& inputs) {
     ret[out_idx] = {outputs_shapes[out_idx], pipeline_.GetOutputType(out_idx),
                     pipeline_.GetOutputDevice(out_idx)};
   }
-  inputs_consumed_ = true; // this will change with introduction of streamed input
+  for (auto &name: input_names_) {
+    auto trace = pipeline_.GetOperatorTrace(name, "next_output_data_id");
+    if (trace != std::to_string(request_id_)) {
```
Comment (JanuszL) [inline src/dali_executor/dali_executor.cc:159]:
```suggestion
    if (std::stoill(trace) != request_id_) {
```
It is a bit more intuitive to compare ints than strings.

Comment (JanuszL) [reply src/dali_executor/dali_executor.cc:159]: Or it won't work if there is no trace available?

File: src/dali_executor/dali_executor.cc [hunk 3]
```diff
@@ -134,7 +138,9 @@ bool DaliExecutor::IsNoCopy(device_type_t es_device, const IDescr& input) {
 }
 
 std::vector<OutputInfo> DaliExecutor::Run(const std::vector<IDescr>& inputs) {
-  SetupInputs(inputs);
+  if (inputs_consumed_) {
+    SetupInputs(inputs);
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_executor.cc:141]: done

File: src/dali_executor/dali_executor.cc (hunk 2 above)
Comment (banasraf) [reply src/dali_executor/dali_executor.cc]: Done

**Review by JanuszL — APPROVED** [review]

File: qa/L0_video_input_decoupled/client.py (hunk 1 above)
Comment (banasraf) [reply qa/L0_video_input_decoupled/client.py:92]: why this?

File: src/utils/triton.h [hunk 4]
```diff
@@ -380,8 +380,11 @@ class TritonResponseView : public TritonResponseWrapper<TritonResponseView> {
   TRITONBACKEND_Response *handle_ = nullptr;
 };
 
-/** @brief Consume and send response and error. */
-void SendResponse(TritonResponse response, TritonError error = TritonError{nullptr});
+/** @brief Consume and send response and error.
+ * final_response - true if it's the last response for a current request
+*/
+void SendResponse(TritonResponse response, bool final_response = true,
+                  TritonError error = TritonError{nullptr});
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/utils/triton.h]: done?

File: qa/L0_video_input_decoupled/model_repository/model.dali/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import nvidia.dali as dali
+import nvidia.dali.fn as fn
+import multiprocessing as mp
+import argparse
+
+@dali.pipeline_def(batch_size=3, num_threads=1, device_id=0, output_ndim=4, output_dtype=dali.types.UINT8)
+def pipeline():
+  return dali.fn.experimental.inputs.video(sequence_length=5, name='INPUT', last_sequence_policy='pad')
```
[the remark this answers is not in the document]
Comment (banasraf) [reply qa/L0_video_input_decoupled/model_repository/model.dali/pipeline.py]: Done

File: src/dali_executor/dali_pipeline.cc [hunk 6]
```diff
@@ -146,4 +151,14 @@ int DaliPipeline::GetMaxBatchSize() {
   return daliGetMaxBatchSize(&handle_);
 }
 
+std::optional<std::string> DaliPipeline::GetOperatorTrace(std::string_view operator_name,
+                                                          std::string_view trace_name) {
+  if (daliHasOperatorTrace(&handle_, operator_name.data(), trace_name.data())) {
+    std::string trace = daliGetOperatorTrace(&handle_, operator_name.data(), trace_name.data());
+    return trace;
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_pipeline.cc]: done

File: src/dali_executor/dali_pipeline.cc [hunk 7]
```diff
@@ -146,4 +151,14 @@ int DaliPipeline::GetMaxBatchSize() {
   return daliGetMaxBatchSize(&handle_);
 }
 
+std::optional<std::string> DaliPipeline::GetOperatorTrace(std::string_view operator_name,
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_pipeline.cc]: done

**Review by szalpal — APPROVED** [review]

**Pull request merged**